scipy/_lib/src/ccallback.h

```diff
@@ -0,0 +1,363 @@
+    }
+
+    capsule = PyDict_GetItemString(local_dict, "__scipy_ccallback");
+    if (capsule == NULL) {
+        return NULL;
+    }
+
+    callback_ptr = PyCapsule_GetPointer(capsule, NULL);
+    if (callback_ptr == NULL) {
+        Py_FatalError("scipy/ccallback: invalid callback state");
+    }
+
+    return callback_ptr;
+}
+
+static int ccallback__set_thread_local(void *value)
+{
+    PyObject *local_dict;
+
+    local_dict = PyThreadState_GetDict();
+    if (local_dict == NULL) {
+        Py_FatalError("scipy/ccallback: failed to get local thread state");
+    }
+
+    if (value == NULL) {
+        return PyDict_DelItemString(local_dict, "__scipy_ccallback");
+    }
+    else {
+        PyObject *capsule;
+        int ret;
+
+        capsule = PyCapsule_New(value, NULL, NULL);
+        if (capsule == NULL) {
+            return -1;
+        }
+        ret = PyDict_SetItemString(local_dict, "__scipy_ccallback", capsule);
+        Py_DECREF(capsule);
+        return ret;
+    }
+}
+
+static ccallback_t *ccallback_obtain(void)
+{
+    PyGILState_STATE state;
+    ccallback_t *callback_ptr;
+
+    state = PyGILState_Ensure();
+
+    callback_ptr = (ccallback_t *)ccallback__get_thread_local();
+    if (callback_ptr == NULL) {
+        Py_FatalError("scipy/ccallback: failed to get thread local state");
+    }
+
+    PyGILState_Release(state);
+
+    return callback_ptr;
+}
+
+#endif
+
+
+/* Set up callback. */
+static int ccallback_prepare(ccallback_t *callback, char **signatures, PyObject *callback_obj, int flags)
+{
+    static PyTypeObject *lowlevelcallable_type = NULL;
+    PyObject *callback_obj2 = NULL;
+    PyObject *capsule = NULL;
+
+    if (lowlevelcallable_type == NULL) {
+        PyObject *module;
+
+        module = PyImport_ImportModule("scipy._lib._ccallback");
+        if (module == NULL) {
+            goto error;
+        }
+
+        lowlevelcallable_type = (PyTypeObject *)PyObject_GetAttrString(module, "LowLevelCallable");
+        if (lowlevelcallable_type == NULL) {
+            goto error;
+        }
+    }
+
+    if ((flags & CCALLBACK_PARSE) && !PyObject_TypeCheck(callback_obj, lowlevelcallable_type)) {
+        /* Parse callback */
+        callback_obj2 = PyObject_CallMethod(lowlevelcallable_type, "_parse_callback",
+                                            "O", callback_obj);
+        if (callback_obj2 == NULL) {
+            goto error;
+        }
+
+        callback_obj = callback_obj2;
+
+        if (PyCapsule_CheckExact(callback_obj)) {
+            capsule = callback_obj;
+        }
+    }
+
+    if (PyCallable_Check(callback_obj)) {
+        /* Python callable */
+        callback->py_function = callback_obj;
+        callback->c_function = NULL;
+        callback->user_data = NULL;
+        callback->signature_index = -1;
+    }
+    else if (PyObject_TypeCheck(callback_obj, lowlevelcallable_type) &&
+             PyTuple_Check(callback_obj) &&
```

Comment:
I'm probably misreading this but... how can `callback_obj` be a tuple and a LowLevelCallable instance at the same time? (unless LowLevelCallable is a tuple subclass)